### PR TITLE
April Fools 2024 easter egg: Scratch Haddocks

### PR DESCRIPTION
**Pull request by WorldLanguages**
Right now only the top part of the popup has any changes at all

**(THIS SCREENSHOT IS NOW OUTDATED)**
image

### CODE REVIEW

File: libraries/common/global-theme.js [hunk 1]
```diff
@@ -6,6 +6,11 @@ export default function () {
     const favicon = document.getElementById("favicon");
     if (favicon) favicon.href = chrome.runtime.getURL("/images/icon-blue.png");
   }
+  const now = Date.now() / 1000;
+  // TODO: update these timestamps for year 2024
+  if (/*now < 1648911600 && now > 1648738800 && */ ["popup"].includes(location.pathname.split("/")[2])) {
+    document.documentElement.style.setProperty("--brand-orange", "#3232fd");
```
Comment (WorldLanguages) [inline libraries/common/global-theme.js:12]: This is not the same blue as the prerelease versions

Comment (OneShot-Niko) [reply libraries/common/global-theme.js:12]: > This is not the same blue as the prerelease versions

Sorry if this is a dumb question that may be a bit off-topic, but, why did you say that in a review on your own PR while you could've just put that in a comment?

**Comment by mybearworld** [comment]
Timestamps for this year are `now < 1712070000 && now > 1711897200`

**Comment by DNin01** [comment]
Is there anything else we could replace with emojis of sea creatures? 😁 Maybe some of the icons?

Or could we just have haddocks swimming/floating around on the popup?

**Comment by DNin01** [comment]
> Or could we just have haddocks swimming/floating around on the popup?

Made it:

https://github.com/ScratchAddons/ScratchAddons/assets/106490990/afe50595-95c9-4fd4-924c-2b952b9a17cb

These are the fish emoji from Windows 11 (the OS I'm using), BTW

**Comment by DNin01** [comment]
Code for that available at https://github.com/DNin01/ScratchAddons/tree/afd24-fish-tank

@WorldLanguages You can add it to this PR if you want.

**Comment by WorldLanguages** [comment]
> Or could we just have haddocks swimming/floating around on the popup?

lol! I don't know what to think

**Comment by WorldLanguages** [comment]
> Timestamps for this year are `now < 1712070000 && now > 1711897200`

I'm not sure if those are 3 hours off or not? Time zones make it confusing
My intention was always to make our April Fools be from March 31 noon UTC+0 to April 2 noon UTC+0

**Comment by Waakul** [comment]
Maybe change the addons tab to Haddocks? That might require lot of translation changes tho. Is that a problem? We could also change the addons wrench icon to a haddock icon.

**Comment by Waakul** [comment]
> > Timestamps for this year are `now < 1712070000 && now > 1711897200`
> 
> I'm not sure if those are 3 hours off or not? Time zones make it confusing My intention was always to make our April Fools be from March 31 noon UTC+0 to April 2 noon UTC+0

Why use timestamps?
```js
var currentDate = new Date();

var currentMonth = currentDate.getMonth();
var currentDay = currentDate.getDate();

if ((currentMonth === 2 && currentDay >= 31) || (currentMonth === 3 && currentDay <= 2)) {
    // The code to do when the date is march 31st - April 2nd here.
}
```
Can do the trick..

**Comment by mybearworld** [comment]
I think we want it to appear at the same time for all users. Timestamps ignore time zones, your code doesn't.

**Comment by Samq64** [comment]
> `currentMonth === 2 && currentDay >= 31`

~~February 31st, what a leap year!~~

**Comment by Waakul** [comment]
> > `currentMonth === 2 && currentDay >= 31`
> 
> ~February 31st, what a leap year!~

Months start from 0 in js if you're wondering.

**Comment by CST1229** [comment]
> Months start from 0 in js if you're confused.

And days don't?!

**Comment by mybearworld** [comment]
> And days don't?!

Welcome to JavaScript!

**Comment by WorldLanguages** [comment]
> > Or could we just have haddocks swimming/floating around on the popup?
> 
> Made it:

Should we add those then? As long as they aren't that annoying, it should be fine...

**Comment by WorldLanguages** [comment]
(We wouldn't show the fish animations if reduced-motion was toggled by the user)

**Comment by BroJac5246** [comment]
I propose that we also temporarily replace the dangos with haddocks because it's easy and why not?

**Comment by WorldLanguages** [comment]
> I propose that we also temporarily replace the dangos with haddocks because it's easy and why not?

Replace *which* dangos?

**Comment by Waakul** [comment]
If we could collect some projects related to haddocks, we could feature them in a separate section under the featured projects section in the homepage like featured dangos?

**Comment by mybearworld** [comment]
> If we could collect some projects related to haddocks, we could feature them in a separate section under the featured projects section in the homepage like featured dangos?

1. Meh, not sure if we want to reuse that
2. I'd be very surprised if there are actual good projects about haddocks lol

**Comment by WorldLanguages** [comment]
What shade of blue is better to represent fish?

**Comment by Waakul** [comment]
> What shade of blue is better to represent fish?

Aqua blue in my opinion.

**Comment by Waakul** [comment]
The logo needs a bit of tweaking to have a fish icon instead of a toolbox one...

**Comment by WorldLanguages** [comment]
chrome_b8yUpNN8kz

File: webpages/popup/index.js [hunk 2]
```diff
@@ -132,3 +138,52 @@ chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
...
+  fishElem.style.position = "absolute";
+  fishElem.style.left = initX + "px";
+  fishElem.style.top = initY + "px";
+  fishElem.style.fontSize = "22px";
+  fishElem.style.textShadow = "#000 0px 2px 5px";
+  new Haddock(document.body.appendChild(fishElem));
+}
+
+const now = Date.now();
+if (true /* TODO */) {
```
Comment (mybearworld) [inline webpages/popup/index.js:184]:
```suggestion
if (!window.matchMedia("(prefers-reduced-motion)").matches /* TODO */) {
```

Comment (WorldLanguages) [reply webpages/popup/index.js:184]: Thanks, I almost forgot

**Comment by mybearworld** [comment]
> I feel like water bubbles should come out of their mouth.

That might be hard to get right because we don't know the specific look of the fish emoji.

**Comment by Waakul** [comment]
> > I feel like water bubbles should come out of their mouth.
> 
> That might be hard to get right because we don't know the specific look of the fish emoji.

We can use svg's instead.

**Comment by mybearworld** [comment]
Could the fish's movement area be a bit smaller? They often go *right* into a corner for me.

**Comment by DNin01** [comment]
> Could the fish's movement area be a bit smaller? They often go *right* into a corner for me.

We might be able to partially remediate this by aligning the text. Could also just subtract the width and height of the element from the max X and Y position.

**Comment by BroJac5246** [comment]
> > I propose that we also temporarily replace the dangos with haddocks because it's easy and why not?
> 
> Replace *which* dangos?

dango-rain. It's sort of odd having a previous year's easter egg alongside this year's.

File: webpages/popup/index.js (hunk 2 above)
Comment (DNin01) [reply webpages/popup/index.js]: If preferring reduced motion makes the swimming fish emojis disappear, then we should probably keep them in the header for those users.

**Comment by DNin01** [comment]
These haddocks can be kind of obtrusive when reading or writing comments, though.
Fish emoji overlaid on comment with reply field

I don't know, maybe we could hide them or make them transparent while you're writing a comment.
Transparent fish emoji overlaid on comment with reply field

**Comment by DNin01** [comment]
Okay, the haddocks shouldn't swim through the edges anymore.

**Comment by RedGuy12** [comment]
Are we still planning to merge this and release 1.37, and expecting CWS to review this before Monday? There's not much time...

File: webpages/popup/index.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ const vue = new Vue({
       const now = Date.now() / 1000;
       if (message === "extensionName" /*&& now < 1648911600 && now > 1648738800*/) {
         // TODO: create whitelist of languages where this change is appropriate
-        return "Scratch Haddocks";
+        return window.matchMedia("(prefers-reduced-motion)").matches ? "Scratch Haddocks 🐟🐟" : "Scratch Haddocks";
```
Comment (WorldLanguages) [inline webpages/popup/index.js:31]: @DNin01 There's not going to be enough space in most screens.

**Comment by WorldLanguages** [comment]
Looks ready to merge @DNin01 @Samq64 @mybearworld Any approvals?

To test it you can replace the 4 instances of `const now = Date.now() / 1000;` with `const now = 1711886401;`

**Comment by WorldLanguages** [comment]
I also tested this with reduced motion enabled.

**Review by mybearworld — APPROVED** [review]
The contrast of the fish in the header (when `prefers-reduced-motion` is enabled) isn't great on Windows 11, but that's not that important.

Scratch Haddocks 🐟🐟🐟 on a blue background where the fish emoji are also blue.

**Comment by Waakul** [comment]
The logo is'nt changed yet...

**Comment by mybearworld** [comment]
I might be wrong on this, but changing the logo would probably delay the process of getting this approved on the web stores further.

**Comment by Waakul** [comment]
Frame (2)
I updated my haddocks logo a bit so the icon is visible

**Comment by mybearworld** [comment]
> I might be wrong on this, but changing the logo would probably delay the process of getting this approved on the web stores further.

Also, changing the logo dynamically only on April Fools sounds like a headache...

**Comment by Waakul** [comment]
> > I might be wrong on this, but changing the logo would probably delay the process of getting this approved on the web stores further.
> 
> Also, changing the logo dynamically only on April Fools sounds like a headache...

By logo i only meant on the popup. See https://github.com/WorldLanguages/ScratchAddons/pull/5

**Comment by WorldLanguages** [comment]
Reviews within 1 hour please

**Review by Samq64 — APPROVED** [review]
Tested on Firefox

**Review by DNin01 — APPROVED** [review]
LGTM

**Comment by mybearworld** [comment]
> These haddocks can be kind of obtrusive when reading or writing comments, though.
> 
> I don't know, maybe we could hide them or make them transparent while you're writing a comment.

That probably should've been implemented - these haddocks are actually pretty annoying when trying to write a comment.

**Comment by DNin01** [comment]
> > These haddocks can be kind of obtrusive when reading or writing comments, though.
> > I don't know, maybe we could hide them or make them transparent while you're writing a comment.
> 
> That probably should've been implemented - these haddocks are actually pretty annoying when trying to write a comment.

Yup...

The good things are:
1. You can always open the comment on Scratch.
2. You can also click the small little icon on the Messaging tab to open the whole Messaging popup in a new tab, which doesn't have haddocks swimming around. Not sure how many people will realize that, though...
3. You could also use a text editor to write your comment without distractions, then paste it in, I suppose.

**Pull request merged**